runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -1179,17 +1179,37 @@ TR_J9ServerVM::isClassVisible(TR_OpaqueClassBlock *sourceClass, TR_OpaqueClassBl
 void *
 TR_J9ServerVM::getJ2IThunk(char *signatureChars, uint32_t signatureLength, TR::Compilation *comp)
    {
-   // Return NULL, since thunk will be recreated on the server whether or not it already exists.
-   // TODO: Track which thunks were created per-client, so that we don't need to recreate them.
+   // Check a set of thunks already registered (or about to be registered) for this client.
+   // If a thunk is there, return -1 to indicate that we don't need to regenerate thunk on
+   // the server and send it to client again. -1 is fine here, because the address will be relocated
+   // on the client anyway.
+   // If a thunk is not registered, return NULL to indicate that we need to generate and register it.
+      {
+      OMR::CriticalSection thunkMonitor(_compInfoPT->getClientData()->getThunkSetMonitor());
+      auto &thunkSet = _compInfoPT->getClientData()->getRegisteredJ2IThunkSet();
+      std::string signature(signatureChars, signatureLength);
+      if (thunkSet.find(signature) != thunkSet.end())
+         return reinterpret_cast<void *>(-1);
```

Comment:
Is it at all possible that this may cause a crash because we provided a fake thunk address and the compiler (at the server) could try to peek inside?